applications/FluidDynamicsApplication/custom_response_functions/velocity_pressure_norm_square_response_function.h

```diff
@@ -0,0 +1,451 @@
+        Vector& rResponseGradient,
+        const ProcessInfo& rProcessInfo) override
+    {
+        CalculateEntityFirstDerivatives<Element>(
+            rAdjointElement, rResidualGradient, rResponseGradient, rProcessInfo);
+    }
+
+    void CalculateFirstDerivativesGradient(
+        const Condition& rAdjointCondition,
+        const Matrix& rResidualGradient,
+        Vector& rResponseGradient,
+        const ProcessInfo& rProcessInfo) override
+    {
+        CalculateEntityFirstDerivatives<Condition>(
+            rAdjointCondition, rResidualGradient, rResponseGradient, rProcessInfo);
+    }
+
+    void CalculateSecondDerivativesGradient(
+        const Element& rAdjointElement,
+        const Matrix& rResidualGradient,
+        Vector& rResponseGradient,
+        const ProcessInfo& rProcessInfo) override
+    {
+        rResponseGradient.clear();
+    }
+
+    void CalculateSecondDerivativesGradient(
+        const Condition& rAdjointCondition,
+        const Matrix& rResidualGradient,
+        Vector& rResponseGradient,
+        const ProcessInfo& rProcessInfo) override
+    {
+        rResponseGradient.clear();
+    }
+
+    void CalculatePartialSensitivity(
+        Element& rAdjointElement,
+        const Variable<array_1d<double, 3>>& rVariable,
+        const Matrix& rSensitivityMatrix,
+        Vector& rSensitivityGradient,
+        const ProcessInfo& rProcessInfo) override
+    {
+        CalculateEntitySensitivityDerivatives(rAdjointElement, rSensitivityGradient, rProcessInfo);
+    }
+
+    void CalculatePartialSensitivity(
+        Condition& rAdjointCondition,
+        const Variable<array_1d<double, 3>>& rVariable,
+        const Matrix& rSensitivityMatrix,
+        Vector& rSensitivityGradient,
+        const ProcessInfo& rProcessInfo) override
+    {
+        CalculateEntitySensitivityDerivatives(rAdjointCondition, rSensitivityGradient, rProcessInfo);
+    }
+
+    double CalculateValue(ModelPart& rModelPart) override
+    {
+        KRATOS_TRY
+
+        double norm_square = 0.0;
+        if (mIsElements) {
+            norm_square += block_for_each<SumReduction<double>>(
+                rModelPart.Elements(), Matrix(),
+                [&](ModelPart::ElementType& rElement, Matrix& rMatrix) -> double {
+                    return CalculateEntityValue<Element>(rElement, rMatrix);
+                });
+        }
+
+        if (mIsConditions) {
+            norm_square += block_for_each<SumReduction<double>>(
+                rModelPart.Conditions(), Matrix(),
+                [&](ModelPart::ConditionType& rCondition, Matrix& rMatrix) -> double {
+                    return CalculateEntityValue<Condition>(rCondition, rMatrix);
+                });
+        }
+
+        return rModelPart.GetCommunicator().GetDataCommunicator().SumAll(norm_square);
+
+        KRATOS_CATCH("");
+    }
+
+    ///@}
+
+private:
+    ///@name Member Variables
+    ///@{
+
+    ModelPart& mrModelPart;
+    std::string mNormModelPartName;
```

Comment:
Why saving both?

Reply:
`mNormModelPartName` holds a submodel part of `mrModelPart`

Reply:
Then I'd save either a reference to that one or a reference to the model together with the name.

Reply:
done :)